src/hato/client.clj

```diff
@@ -284,42 +275,114 @@
                       (->BodyPublisher req))
              .build)))
 
+(defn unwrap-async-exception
+  "Helper function to unwrap common async exceptions to get the root cause such as
+  an IOException."
+  [^Exception ex]
+  (if (or (instance? CompletionException ex)
+          (instance? ExecutionException ex))
+    (if-let [cause (.getCause ex)]
+      (unwrap-async-exception cause)
+      ex)
+    ex))
+
+(def retry-exceptions
+  "Common exceptions to retry requests on."
+  #{IOException})
+
+(def non-retry-exceptions
+  "Common exceptions to not retry requests on. Borrowed from Apache HttpClient's
+  `DefaultHttpRequestRetryHandler`."
+  #{InterruptedIOException UnknownHostException SSLException})
+
+(defn idempotent-request?
+  "Checks to see if the request is idempotent. It is considered idempotent if the
+  request method is one of `:get`, `:head`, or `:options`."
+  [req]
+  (contains? #{:get :head :options} (:request-method req)))
+
+(defn make-retry-handler
+  "Constructs a basic retry handler. Options are:
+
+  :retry-exceptions     => A collection of exceptions required for us to retry the request on.
+                           Defaults to `retry-exceptions`.
+  :non-retry-exceptions => A collection of exceptions to cause us to skip retrying a specific
+                           request. Defaults to `non-retry-exceptions`.
+  :max-retry-count      => The maximum number of retries before giving up. Defaults to 3."
+  [{:keys [retry-exceptions non-retry-exceptions max-retry-count]
+    :or   {retry-exceptions     retry-exceptions
+           non-retry-exceptions non-retry-exceptions
+           max-retry-count      3}}]
+  (fn [_resp ex req retry-count]
+    (and (idempotent-request? req)
+         (some #(instance? % ex) retry-exceptions)
+         (not (some #(instance? % ex) non-retry-exceptions))
+         (< retry-count max-retry-count))))
+
+(def default-retry-handler
+  "A default retry handler which will retry if there is an error with any known retriaretryble
+  exceptions while skipping non-retry exceptions. By default, this will only retry idempotent
+  requests (i.e. GET/HEAD/OPTIONS)."
+  (make-retry-handler {}))
+
+(defn request-with-retries*
+  "Makes a request retrying on failures as defined by retry-handler"
+  [http-client http-request body-handler req retry-handler retry-count]
+  (-> (.sendAsync http-client http-request body-handler)
+      (.handle
+        (reify BiFunction
+          (apply [_ resp e]
+            [resp (unwrap-async-exception e)])))
+      (.thenCompose
+        (reify Function
+          (apply [_ [resp e]]
+            (let [retry-fut (if retry-handler
+                              (retry-handler resp e req retry-count)
```

Comment:
Do you think we should put some of this in a map instead so it is more easily extensible in the future? Something like a retry-context map?